jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java

```diff
@@ -14,4 +23,89 @@ public class JdbcTemplate {
     public JdbcTemplate(final DataSource dataSource) {
         this.dataSource = dataSource;
     }
+
+    public int update(String sql, Object... args) {
+        return execute(sql, pstmt -> {
+            setParameters(pstmt, args);
+            return pstmt.executeUpdate();
+        });
+    }
+
+    public <T> T queryForObject(String sql, RowMapper<T> rowMapper, Object... args) {
+        List<T> results = query(sql, rowMapper, args);
+        
+        if (results.isEmpty()) {
+            throw new EmptyResultDataAccessException("Query returned no results");
+        }
+        
+        if (results.size() > 1) {
+            throw new IncorrectResultSizeDataAccessException(1, results.size());
+        }
+        
+        return results.getFirst();
+    }
+
+    public <T> List<T> query(String sql, RowMapper<T> rowMapper, Object... args) {
+        return execute(sql, pstmt -> {
+            setParameters(pstmt, args);
+            
+            try (ResultSet rs = pstmt.executeQuery()) {
+                List<T> results = new ArrayList<>();
+                int rowNum = 0;
+                
+                while (rs.next()) {
+                    results.add(rowMapper.mapRow(rs, rowNum++));
+                }
+                
+                return results;
+            }
+        });
+    }
+
+    private <T> T execute(String sql, PreparedStatementCallback<T> callback) {
+        Connection conn = null;
+        PreparedStatement pstmt = null;
+        
+        try {
+            conn = dataSource.getConnection();
+            pstmt = conn.prepareStatement(sql);
+            
+            log.debug("Executing SQL: {}", sql);
+            
+            return callback.doInPreparedStatement(pstmt);
+            
+        } catch (SQLException e) {
+            log.error("SQL execution failed: {}", sql, e);
+            throw new DataAccessException("SQL execution failed: " + sql, e);
+        } finally {
+            closeStatement(pstmt);
+            closeConnection(conn);
+        }
+    }
```

Comment:
```suggestion
private <T> T execute(String sql, PreparedStatementCallback<T> callback) {
    try (Connection conn = dataSource.getConnection();
         PreparedStatement pstmt = conn.prepareStatement(sql)) {
        
        log.debug("Executing SQL: {}", sql);
        return callback.doInPreparedStatement(pstmt);
    } catch (SQLException e) {
        log.error("SQL execution failed: {}", sql, e);
        throw new DataAccessException("SQL execution failed: " + sql, e);
    }
}
```

이렇게 try-with-resources 구문을 활용하면 더 안전하게 리소스를 close하고 불필요한 close 메서드 구현을 할 필요가 없을것 같아요. 어떤 경우 `null`이 들어오는 것을 고려해 따로 close 메서드를 구현했을까요??

Reply:
@ykmxxi 
오 맞네요!

null 체크를 한 이유는 아래와 같아요.
dataSource.getConnection()에서 예외 발생 시 conn이 null
conn.prepareStatement(sql)에서 예외 발생 시 pstmt가 null

이런 경우에 finally 블록에서 null.close()를 호출하면 NPE가 발생하기 때문에 null 체크가 필요했습니다.
하지만 뽀스티 말처럼 try-with-resources를 사용하면 더 좋을 것 같네요!

살짝 고민되는 건 트랜잭션을 고려했을 때, 기존 트랜잭션이 존재 시에는 그때 사용하던 Connection을 재사용해야하고 메서드가 종료되어도 Connection이 닫히면 안돼서 그땐 try-with-resources를 사용하기 힘들 것 같은데 그건 그때가서 생각해보겠습니다!

뽀스티가 제안해준 방식으로 step2에 수정해볼게요!